includes/sensei-functions.php

```diff
@@ -38,25 +38,45 @@ function is_sensei() {
 }
 
 /**
- * Determine if the current user is and admin that
- * can acess all of Sensei without restrictions
+ * Determine if a user is an admin that can access all of Sensei without restrictions.
  *
  * @since 1.4.0
+ * @since 3.0.0 Added `$user_id` argument. Preserves backward compatibility.
+ *
+ * @param int $user_id User ID. Defaults to current user.
+ *
  * @return boolean
  */
-function sensei_all_access() {
+function sensei_all_access( $user_id = null ) {
+	if ( null === $user_id ) {
+		$user_id = get_current_user_id();
+	}
+
+	if ( empty( $user_id ) ) {
+		return false;
+	}
+
+	$access = user_can( $user_id, 'manage_sensei' ) || user_can( $user_id, 'manage_sensei_grades' );
 
-	$access = current_user_can( 'manage_sensei' ) || current_user_can( 'manage_sensei_grades' );
+	// For backwards compatibility with filter, we temporarily need to change the current user.
+	$previous_current_user_id = get_current_user_id();
+	wp_set_current_user( $user_id );
```

Comment:
Something about this seems scary to me. I understand the purpose, I just wonder if there's something weird that the `sensei_all_access` filters could do because of this... I don't have anything concrete though, so it might be an unnecessary worry 🙂

Reply:
Generally this is only used with the current user so these methods don't change anything. Alternatively we could:
- Deprecate the old hook and only do this jigging when the old hook is set.
- Create a new but very similarly named hook :). (open to suggestions if you want to go down this route).

Reply:
I'd say any hook handler that is _doing_ something on this filter is bad anyway (and doing bad things already) so my gut is we just consider the usages in the plugin itself.

Reply:
Yeah, I think we can safely consider side-effects from handlers to be unsupported behaviour, since this is a filter, not an action.

I wonder, though, if it would be good to deprecate the old functionality (i.e. the usage of the current user instead of the passed-in `$user_id`)? Otherwise I feel like this is going to be a long-lived "hack" for backwards compatibility, and it would be nice to be able to get rid of it eventually.

But I'm not sure how to do that other than deprecating the filter and creating a new one (and changing all of our internal usages of it), and I don't think it's a big enough problem to warrant that amount of work/testing.

WDYT?

Reply:
It doesn't seem too dangerous to try the deprecated hook route: 4d8ae77968ffa87b43c1d86b8799854a8ba16886

How does that look?

Reply:
Nice! Yeah that looks great. I didn't realize that we aren't actually using that filter at all internally, but I searched and didn't find anything 👍